src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py

```diff
@@ -307,46 +309,58 @@ def _dispatchWflow(self, wflow):
             # NOTE: We skip workflows which have not yet finalised their TransferStatus
             #       in MSOutput, but we still need some proper logging for them.
-            msg = "Skipping workflow: %s - 'TransferStatus' is 'pending' or 'TransferInfo' is missing in MSOutput."
+            msg = "Skipping workflow: %s - 'TransferStatus' is 'pending' or 'TransferInfo' is missing in MSOutput." % wflow['RequestName']
             msg += " Will retry again in the next cycle."
-            self.logger.info(msg, wflow['RequestName'])
+            self.logger.info(msg)
+            if self._isStatusAdvanceExpired(wflow, wflow['RequestStatus']):
+                self.alertStatusAdvanceExpired(wflow, msg)
         elif wflow['RequestStatus'] == 'announced' and not wflow['TransferTape']:
             # NOTE: We skip workflows which have not yet finalised their tape transfers.
             #       (i.e. even if a single output which is supposed to be covered
             #       by a tape rule is in any of the following transient states:
             #       {REPLICATING, STUCK, SUSPENDED, WAITING_APPROVAL}.)
             #       We still need some proper logging for them.
-            msg = "Skipping workflow: %s - tape transfers are not yet completed."
+            msg = "Skipping workflow: %s - tape transfers are not yet completed." % wflow['RequestName']
             msg += " Will retry again in the next cycle."
-            self.logger.info(msg, wflow['RequestName'])
+            self.logger.info(msg)
+            if self._isStatusAdvanceExpired(wflow, wflow['RequestStatus']):
+                self.alertStatusAdvanceExpired(wflow, msg)
         elif wflow['RequestStatus'] == 'announced':
             for pline in self.cleanuplines:
                 try:
                     pline.run(wflow)
                 except MSRuleCleanerResolveParentError as ex:
-                    msg = "%s: Parentage Resolve Error: %s. "
+                    msg = "%s: Parentage Resolve Error: %s. " % (pline.name, str(ex))
                     msg += "Will retry again in the next cycle."
-                    self.logger.error(msg, pline.name, str(ex))
+                    self.logger.error(msg)
+                    if self._isStatusAdvanceExpired(wflow, wflow['RequestStatus']):
+                        self.alertStatusAdvanceExpired(wflow, msg)
                     continue
                 except Exception as ex:
-                    msg = "%s: General error from pipeline. Workflow: %s. Error:  \n%s. "
+                    msg = "%s: General error from pipeline. Workflow: %s. Error:  \n%s. " % (pline.name, wflow['RequestName'], str(ex))
                     msg += "\nWill retry again in the next cycle."
-                    self.logger.exception(msg, pline.name, wflow['RequestName'], str(ex))
+                    self.logger.exception(msg)
+                    if self._isStatusAdvanceExpired(wflow, wflow['RequestStatus']):
```

Comment:
I'd suggest to move this check out of a generic exception and instead catch those custom exceptions (ArchivalSkip, ArchivalError, etc). You probably don't want to through the stack trace in the log for all those exceptions as well (self.logger.exception). In other words, update this try/except/except to consider those exceptions and give them a different treatment compared to this generic Exception.

Reply:
Alan, I do not get what you mean by this. If it is that  you prefer not to issue any alarms when we  have a workflow that  is constantly being dropped because of hitting some unknown in advance bug in the cleaning process, then I am ok. I
ll remove the alarm from here. 

As of:
> In other words, update this try/except/except to consider those exceptions and give them a different treatment compared to this generic Exception.

Please elaborate. What exactly do you mean here?

Reply:
I wanted to say that this code block:
```
                try:
                    pline.run(wflow)
                except MSRuleCleanerResolveParentError as ex:
                    msg = "%s: Parentage Resolve Error: %s. " etc etc
                except Exception as ex:
                    msg = "%s: General error from pipeline. Workflow: %s. Error:  \n%s. " etc etc
```

should probably be something like (note the new and expected/possible exceptions):
```
                try:
                    pline.run(wflow)
                except MSRuleCleanerResolveParentError as ex:
                    msg = "%s: Parentage Resolve Error: %s. " etc etc
                except (MSRuleCleanerArchivalSkip, MSRuleCleanerArchivalError, etc) as ex:
                    msg = "%s: Known error blah blah blah: %s. " etc etc
                    self.logger.error
                except Exception as ex:
                    msg = "%s: General error from pipeline. Workflow: %s. Error:  \n%s. " etc etc
                    self.logger.exception
```

But please review which exceptions can be raised and caught here.

With this construction, I think that most - if not all - of the workflows will fall under one of the constraint-based custom exceptions, instead of any unknown error that would be trapped under the generic Exception.

But yes, if you think we can have workflows hitting a general exception over and over, then we might want to make an alert for them as well.

Reply:
[Alan](@amaltaro) , at this stage none of those exceptions could be raised. This is possible only on the later step during archival, and all relevant cases related to them are already handled with proper alerts.